[mono][eventpipe] Implementing MethodDetails and BulkType events

mdh1418 opened the pull request:
This PR looks to implement MethodDetails and BulkType trace events on mono, closely following the implementation of src/coreclr/vm/eventtrace.cpp's [SendMethodDetailsEvent](https://github.com/dotnet/runtime/blob/e4163ea55ebb3673c29e1c2a850a6a790029d278/src/coreclr/vm/eventtrace.cpp#L6413) and [FireBulkTypeEvent](https://github.com/dotnet/runtime/blob/e4163ea55ebb3673c29e1c2a850a6a790029d278/src/coreclr/vm/eventtrace.cpp#L1641). In doing so, apps running on mono can be traced following steps from https://github.com/dotnet/runtime/blob/main/docs/design/mono/diagnostics-tracing.md and produce a `.nettrace` file with relevant MethodDetails and BulkType events. This `.nettrace` containing `MethodDetails` and `BulkType` events is necessary to leverage the dotnet-pgo tool to generate a corresponding .mibc file for future consumption in the mono AOT compiler.

This PR makes the following changes:
- Implements MethodDetails events in ep_rt_mono_send_method_details_event (2nd commit)
- Implements BulkType events in ep_rt_mono_fire_bulk_type_event (4th commit)
- Extends LttngDataTypeMapping to have a mono friendly mapping (1st commit)
- Utilizes mono method signature getter to avoid accessing method signatures prior to creation (3rd commit)

----------------
### Creating a `.mibc` on mono from the [HelloWorld sample](https://github.com/dotnet/runtime/blob/main/src/mono/sample/HelloWorld/Program.cs)

1. Building mono+libs
2. Building the HelloWorld sample with `make run`
3. Generating a trace
   3a. `DOTNET_DiagnosticPorts="/Users/mdhwang/myport,suspend" <path to app executable>`
   3b. `dotnet-trace collect --providers Microsoft-Windows-DotNETRuntime:0x1F000080018:5 --diagnostic-port ~/myport`
4. `dotnet-pgo create-mibc --trace <path-to-trace>` <s>(built with a dotnet-pgo tool with [AddAssembliesAssociatedWithType](https://github.com/dotnet/runtime/blob/e4163ea55ebb3673c29e1c2a850a6a790029d278/src/coreclr/tools/dotnet-pgo/MibcEmitter.cs#L170) modified to skip generic parameters)
```
if (type.IsGenericParameter)
  return;
```
</s> Modification no longer needed.

--------------
Implementation Details
-----
For the most part, the implementation of the relevant functions are similar to CoreCLR. Below are some key differences

### `ep_rt_mono_send_method_details_event`
Utilizes a helper `get_typeid_for_type` in order to obtain a unique type ID, to fulfill the niche of CoreCLR's type handle 

### `ep_rt_mono_log_type_and_parameters_if_necessary`
Decided to not implement the hash table until it will actually be used, which seems to be when `ep_rt_mono_log_type_and_parameters_if_necessary` is used outside of `ep_rt_mono_send_method_details_event`

### `ep_rt_mono_log_type_and_parameters`
In line with `LogTypeAndParameters` except for the absence of `typeLogBehavior`, for currently this is only called following `ep_rt_mono_send_method_details_event` which always logs.

### `ep_rt_mono_log_single_type`
Fills in `rg_type_parameters` and `c_type_parameters` by three categories of MonoType  1) `MONO_TYPE_ARRAY` `MONO_TYPE_SZARRAY  2) `MONO_TYPE_GENERICINST`  3) `MONO_TYPE_CLASS` `MONO_TYPE_VALUETYPE` `MONO_TYPE_PTR` `MONO_TYPE_BYREF`  for any embedded type parameters.
Does not set `s_name` and keeps it as `NULL` until the data will actually be used in mono, which GC Heap Dumps are a likely candidate for.

### `ep_rt_mono_fire_bulk_type_event`
Copies to the BulkTypeEvent buffer through helper functions.

Various helper functions and `static const` variables have been added in addition to `BulkTypeValue` and `BulkTypeEventLogger` structs with respective memory management `init/fini/clear` functions.

--------------
TODO/Questions
-----
<s>Does sName in eventtrace.cpp need to be carried over? </s> 
We have opted to keep `s_name` as `NULL` until it is actually used in mono, which GC Heap Dumps are a likely candidate for.

<s> Variable size array for rg_type_parameters in BulkTypeValue struct that already has variable size field s_name </s> 
Figured out that `StackSArray<ULONGLONG>` has a fixed size that evaluates to `32`

Where else to emit ep_rt_mono_send_method_details_event?

<s> How necessary are the 0x06 method token mask and 0x02 type token masks, should dotnet-pgo be tweaked instead? </s> 
Found the equivalent for mono using `mono_metadata_make_token` and `mono_metadata_token_index`

<s> How to implement `if_necessary` portion of LogTypeAndParametersIfNecessary (th.IsRestored and what exactly is client usage of LogTypeAndParametersIfNecessary) </s> 
Decided to not implement the hash table until it will actually be used, which seems to be when `ep_rt_mono_log_type_and_parameters_if_necessary` is used outside of `ep_rt_mono_send_method_details_event`

mdh1418 committed:
[coreclr]Extend lttngDataTypeMapping for mono

Writing arrays in generateWriteEventBody had not been previously hit on Mono.
With a goal of emitting MethodDetails events on Mono, lttngDataTypeMapping
needs to be extended to be compatible with mono types.

mdh1418 committed:
[mono][eventpipe] Implement SendMethodDetailsEvents on mono

mdh1418 committed:
[mono][eventpipe] Implement BulkType on mono

mdh1418 committed:
[mono][eventpipe] Utilize method signature getter to avoid uncreated signature

mdh1418 committed:
debug traces

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		{
+			// Add generic inst parameter's unique type identifier to rg_type_parameters array
+			MonoClass *mono_class_inst_type_class = mono_class_from_mono_type_internal (class_inst->type_argv[i]);
+			if (m_class_is_byreflike (mono_class_inst_type_class))
```
again, this should be `class_inst->type_argv[i]->byref` not `m_class_is_byreflike(...)`

mdh1418 replied:
Done, using the helper function.

lambdageek reviewed (changes requested):
Good start.  Could be a little bit less copy-pasty.

The biggest issue is the use of `m_class_is_byreflike` when you mean to look at `MonoType:byref`.

Also I'm concerned about locking and generally accessing the bulk types buffer from multiple threads.  If multiple threads are writing and any one of them could fire the buffer, there's no guarantee that the other threads finished populating their entry.

I would suggest using a simple mutex to only allow one function to access the buffers at a time (basically make `ep_rt_mono_log_type_and_parameters` take a mutex),

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+uint32_t m_n_bulk_type_value_count;
+uint32_t m_n_bulk_type_value_byte_count;
+uint8_t *m_p_bulk_type_event_buffer[65536];
```
I think this should really be kept as local variables for each method detail events, bulk type emits all its types before sending the method details events, if you are worried about multiple reallocations of the buffer when adding bulk types, you could start with a local stack buffer (that covers most of the cases in its default size) and only if that is not enough move over to dynamic allocated buffer, but should not be needed frequently, unless the method uses a bunch of unique types.

mdh1418 replied:
> also what do the prefixes mean?
I'm not actually sure, I was trying to keep it in parallel with CoreCLR. https://github.com/dotnet/runtime/blob/7d1836ac70233c62d71e2e8339eb9bf4a96457d9/src/coreclr/vm/eventtracepriv.h#L285-L289 but we can rename it if thats preferred.

Originally I just brought the variables in as globals to get things rolling. Thanks for reminding me of threading concerns.

Would we prefer to keep these as local variables for each method details events (would this look like passing around these as parameters?) or mimicking CoreCLR's BulkTypeEventLogger instance with a struct? I initially skipped attempting to implement CoreCLR's BulkTypeEventLogger instance, `TypeSystemLog`, `MethodLog` as I'm not very familiar with (C/C++) and just wanted to get the implementation going. I haven't grasped the broader intent with CoreCLR's structure, but is the current structure in mono roughly sufficient? Would it be better to make instances of each instead?

lateralusX replied:
Since we might end up using the bulk type event logger in more places going forward, I would do something similar as we done with other event pipe types, represent them as a struct and potentially add getter/setters init/fini and alloc/free (if you need heap allocation). Since this type is currently only used in Mono's EP implementation you should define this type within Mono specific EP sources, like ep-rt-mono.h

mdh1418 replied:
Added structs for `BulkTypeEventLogger` that has `ep_rt_bulk_type_event_logger_init` and `ep_rt_bulk_type_event_logger_fini`, and `ep_rt_bulk_type_value_init` for the BulkTypeValues.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+					NULL,
+					NULL);
+
+	m_n_bulk_type_value_count = 0;
+	m_n_bulk_type_value_byte_count = 0;
```
This all should really be using a locals for this, don't see why it needs to be shared between threads, since each set of events are emitted together with the send method details events.

mdh1418 replied:
Added struct local instances that are passed through as pointers.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	p_val->fixed_sized_data.module_id = (uint64_t)m_class_get_image (klass);
+	p_val->fixed_sized_data.type_name_id = (0x00FFFFFF & m_class_get_type_token (klass)) | 0x02000000;  // dotnet-pgo ResolveMethodID expects type name ids with a 0x02 mask
+	p_val->fixed_sized_data.flags = 0;
+	p_val->fixed_sized_data.cor_element_type = (uint8_t)mono_type->type;
```
This must match CoreCLR's logic, TypeHandle::GetInternalCorElementType(), it does some special handling of some types and enums as well.

mdh1418 replied:
I initially saw that TypeDesc's GetInternalCorElementType returns a [`CorElementType`](https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/coreclr/inc/corhdr.h#L870-L923) which is mostly in line with [`MonoTypeEnum`](https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/native/public/mono/metadata/blob.h#L12-L51) except for their `ELEMENT_TYPE_MAX 0x22` and our `MONO_TYPE_ENUM 0x55`. Looks like `MethodTable`'s `GetInternalCorElementType` returns the same type but based off of a flag https://github.com/dotnet/runtime/blob/694067c12626668074197d2f85f7babe6f01ac7a/src/coreclr/vm/methodtable.cpp#L4865-L4917. `CorElementType` and `MonoTypeEnum` *seem* to be equivalent, but I dont know what would be mono's `ELEMENT_TYPE_MAX` or how CoreCLR handles enums like `MONO_TYPE_ENUM`.

lateralusX replied:
I guess this is mostly accurate, but looking the comments in TypeHandle::GetInternalCorElementType made me realize that we might need to see how that applies to Mono as well:

//    * Enumerated type have been normalized to the primitive type that underlies them (typically int)
//    * Value types that look like ints (which include RuntimeTypeHandles, etc), have been morphed to be
//        their underlying type (much like enumeration types.  See

lambdageek replied:
we have some `underlying_type` helpers in various places to handle enums.   In other places where we do a `switch` on `MonoType:type` there's usually a `handle_enum:` label and when we see `MONO_TYPE_VALUETYPE` we get the underlying type of the enum and do a `goto handle_enum`.

For example here:
https://github.com/dotnet/runtime/blob/729bf92e6e2f91aa337da9459bef079b14a0bf34/src/mono/mono/mini/aot-runtime-wasm.c#L22-L53

lambdageek replied:
Also `MONO_TYPE_ENUM 0x55` is used during metadata parsing (it's a marker in signatures I think... I don't remember and don't have my ECMA pdf handy), but we never have a `MonoType:type` of `MONO_TYPE_ENUM`

mdh1418 replied:
Used `mono_type_get_underlying_type (mono_type)` to get any underlying type of enumerated types or value types.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+typedef struct _BultTypeValue {
+	EventStructBulkTypeFixedSizedData fixed_sized_data;
+	uint32_t c_type_parameters;
+	intptr_t rg_type_parameters[100]; // Should be variable length to save space
+	char *s_name;
```
CoreCLR uses a prealloced stack array for this, `StackSArray<ULONGLONG>`.

mdh1418 replied:
When I first looked at it, I wasn't sure what the size of CoreCLR's `StackSArray<ULONGLONG>` was, and naively thought it was an array that can allocate more space on demand. Now with another take, I see that https://github.com/dotnet/runtime/blob/6af4abc5a0db0560e719539a869233d15d28dc09/src/coreclr/inc/sarray.h#L218 has `STACK_ALLOC/sizeof(ELEMENT)` which in this case would be `STACK_ALLOC/sizeof(uint64_t)` and by https://github.com/dotnet/runtime/blob/dbe06f0ce86455149ab140b383f788f9ec110f96/src/coreclr/inc/sbuffer.h#L557, this would be `256/8`.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+				p_val->rg_type_parameters[i] = m_class_get_byval_arg (mono_class_inst_type_class);
+		}
+	}
+
```
Looks like CoreCLR has some more logic around type descriptor, method table (if its a string/object), if type has finalizers, if its a delegate or COM object

mdh1418 replied:
I did notice those, but wasn't sure what the equivalent for `TypeDesc` and `MethodTable` were in mono. What other types in Mono have type parameters besides Arrays and Generic Instantiations?

I'm guessing that TypeHandle.GetInternalCorElementType doesn't resolve to a string or object if its a method table, based on those subcases. If we can pass `mono_type->type` for `p_val->fixed_sized_data.cor_element_type`, would we run into the same scenario where a certain `mono_type` that is a string or object doesn't actually have `mono_type->type` as string or object?

For the Finalizers, delegate, or COM object, do we have similar flags set in mono? Would they happen to be the `MonoType` `attrs`?

lateralusX replied:
mono_class_is_com_object
m_class_is_delegate
mono_class_has_finalizer

> I'm guessing that TypeHandle.GetInternalCorElementType doesn't resolve to a string or object if its a method table, based on those subcases. If we can pass `mono_type->type` for `p_val->fixed_sized_data.cor_element_type`, would we run into the same scenario where a certain `mono_type` that is a string or object doesn't actually have `mono_type->type` as string or object?

I think they will be first encoded as ELEMENT_TYPE_CLASS and then later specialized by the logic checking for string or object.

mdh1418 replied:
Added the corresponding flags and their const to `EtwTypeFlags`

mdh1418 committed:
Address struct tab formatting feedback

mdh1418 committed:
Address rg_type_parameter preallocated stack array size

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+// the bulk type event.  These fields must still match exactly the initial
+// fields of the struct described in the manifest.
+typedef struct _EventStructBulkTypeFixedSizedData {
+	uint64_t type_id;
+    uint64_t module_id;
+    uint32_t type_name_id;
+    uint32_t flags;
+	uint8_t cor_element_type;
+} EventStructBulkTypeFixedSizedData;
```
Thanks!

mdh1418 commented on src/coreclr/scripts/genLttngProvider.py:
```diff
@@ -88,6 +88,34 @@
+        "win:UnicodeString" :"const uint8_t*",
+        "win:Double"        :"const double",
+        "win:Int32"         :"const int32_t",
+        "win:Boolean"       :"const bool",
```
So is it fine to keep it as `bool`? Is the concern about the size? Does `bool` not have a known width?

lateralusX replied:
It will be converted to a 4 byte integer before serialized into the nettrace file format, so it should be fine.

mdh1418 committed:
[mono][eventpipe] Add helper function to get mono type unique type identifier

mdh1418 committed:
[mono][eventpipe] Differentiate MonoType and TypeID and pass separately

mdh1418 committed:
Address array max rank feedback

mdh1418 committed:
Address byte_count bug feedback

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			// Fortunately kEtwTypeFlagsArrayRankMax should be greater than the
+			// number of ranks the type loader will support
+			uint32_t rank = mono_array_type->rank;
+			if (rank < (0x3700 >> 8))
```
Thanks, I just noticed I had a typo as well.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	return
+		sizeof(bulk_type_value->fixed_sized_data) +
+		sizeof(bulk_type_value->c_type_parameters) +
+		(sizeof(bulk_type_value->s_name) + 1) * sizeof(char) +  // Size of name, including null terminator
```
Thanks for the catch :)

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	BulkTypeValue *p_val = &m_rg_bulk_type_values[m_n_bulk_type_value_count];
+
+	MonoType *mono_type = (MonoType*)type_id;
+	MonoClass *klass = mono_class_from_mono_type_internal (mono_type);
```
Done, I had to add an additional parallel array in BulkTypeValue to contain the MonoType associated with the TypeID to make the recursion in `ep_rt_mono_log_type_and_parameters` work.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			if (m_class_is_byreflike (mono_class_inst_type_class))
+				p_val->rg_type_parameters[i] = m_class_get_this_arg (mono_class_inst_type_class);
+			else
+				p_val->rg_type_parameters[i] = m_class_get_byval_arg (mono_class_inst_type_class);
```
Done

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,383 @@ ep_rt_mono_write_event_ee_startup_start (void)
+//      * type_id - Unique type identifier to batch
+
+void
+ep_rt_mono_log_type_and_parameters (intptr_t type_id)
```
Done

mdh1418 committed:
Address Etw Type Flags feedback

mdh1418 committed:
Address cor_element_type underlying type feedback

mdh1418 committed:
Fix s_name

mdh1418 committed:
Cleanup comment tab spacing

mdh1418 committed:
Address logger instance and if_necessary feedback

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// associated with arrays or generics and add unique identifiers to rg_type_parameters array
+	// to be recursively batched in the same ep_rt_mono_log_type_and_parameters call
+	if ((mono_underlying_type == MONO_TYPE_ARRAY) || (mono_underlying_type == MONO_TYPE_SZARRAY))
+	{
+		MonoArrayType *mono_array_type = mono_type_get_array_type (mono_type);
+		p_val->fixed_sized_data.flags |= K_ETW_TYPE_FLAGS_ARRAY;
+		if (mono_underlying_type == MONO_TYPE_ARRAY)
+		{
+			// Only ranks less than kEtwTypeFlagsArrayRankMax are supported.
+			// Fortunately kEtwTypeFlagsArrayRankMax should be greater than the
+			// number of ranks the type loader will support
+			uint32_t rank = mono_array_type->rank;
+			if (rank < K_ETW_TYPE_FLAGS_ARRAY_RANK_MAX)
+			{
+				rank <<= 8;
+				p_val->fixed_sized_data.flags |= rank;
+			}
+		}
+
+		// mono arrays are always arrays of by value types
+		p_val->rg_mono_type_parameters[p_val->c_type_parameters] = m_class_get_byval_arg (mono_array_type->eklass);
+		p_val->rg_type_parameters[p_val->c_type_parameters] = get_typeid_for_type (m_class_get_byval_arg (mono_array_type->eklass));
+		p_val->c_type_parameters++;
+	}
+	else if (mono_underlying_type == MONO_TYPE_GENERICINST)
+	{
+		MonoGenericInst *class_inst = mono_type->data.generic_class->context.class_inst;
+		p_val->c_type_parameters = class_inst->type_argc;
+		for (int i = 0; i < class_inst->type_argc; i++)
+		{
+			p_val->rg_mono_type_parameters[i] = class_inst->type_argv[i];
+			p_val->rg_type_parameters[i] = get_typeid_for_type (class_inst->type_argv[i]);
+		}
+	}
+	else if (mono_underlying_type == MONO_TYPE_CLASS || mono_underlying_type == MONO_TYPE_VALUETYPE || mono_underlying_type == MONO_TYPE_PTR || mono_underlying_type == MONO_TYPE_BYREF)
+	{
+		p_val->rg_mono_type_parameters[p_val->c_type_parameters] = mono_type_get_underlying_type (mono_type);
+		p_val->rg_type_parameters[p_val->c_type_parameters] = get_typeid_for_type (mono_type_get_underlying_type (mono_type));
+		p_val->c_type_parameters++;
+	}
```
nit: `switch` statement?

mdh1418 replied:
Switched to switch :)

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			p_val->s_name[0] = '\0';
+		cb_val = ep_rt_mono_get_byte_count_in_event(p_val);
+
+		if (cb_val > k_max_bytes_type_values)
+		{
+			// This type is apparently so huge, it's too big to squeeze into an event, even
+			// if it were the only type batched in the whole event.  Bail
+			return -1;
+		}
```
This seems like something we may to log occurring to somewhere.

mdh1418 replied:
Logged an error via mono_trace

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	}
+	if (p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)
+	{
+		// Although this type fits into the array, its size is so big that the entire
```
In addition to this comment, can we put an assert or other debug time mechanisms for validating this assertion?

mdh1418 replied:
What exactly are we asserting? Just that the if statement holds? `EP_ASSERT(p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)`? Is this so that in debug mode we know that this was hit?

lateralusX replied:
The statement in the assert expression needs to always be true, or you will get an assert fired that will break the process with a message (the failing expression). EP_ASSERT is mapped to a g_assert_checked on Mono so it's only in checked build.

mdh1418 replied:
Added a g_assert, but would it be more helpful to have an assert for the opposite `p_type_logger->m_n_bulk_type_value_byte_count + cb_val <= k_max_bytes_type_values` after this if?

josalem commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// local copy of their type handles first (else, as we log them we could flush
+	// and clear out m_rg_bulk_type_values, thus trashing p_val)
+	uint32_t c_params = p_val->c_type_parameters;
+	intptr_t rg_type_parameters[c_params];
+	MonoType *rg_mono_type_parameters[c_params];
```
I'm surprised this didn't flag a compiler error (maybe it did; I didn't look at the build logs). You'll want to just use the `const` max number of param types since this is stack allocated.

lambdageek replied:
variable length arrays work in C99 (and on compilers we support).  But yea, there's no real reason to use them here.

mdh1418 replied:
Changed to fixed length array with size `k_size_of_type_parameter_array` (32)

josalem reviewed (changes requested):
A few code cleanliness nits, but the structure is headed in the right direction.

My big worry is there are some pointer manipulations that are incorrect but are not causing issues right now due to another pointer manipulation issue that accidentally corrects it.

mdh1418 committed:
Address feedback

Fix bulk type event buffer

Fix mono type check

Fix mono type check

Clear bulk type event logger

Debug bulktype event firing

Fix tab bulktype firing

Fix fixed_sized_data byte copy

Address constants feedback

Add maximum event size constant
Add maximum bulk type value type parameters constant
Add maximum method type argument count constant

Fix typo

Address BulkType Value s_name type size feedback

Address softcode buffer size increment

Address switch case feedback

Address fixed sized array feedback

Fix spacing

Add init/fini for BulkTypeEventLogger and BulkTypeValue instances

Fix return type ep_rt_mono_log_single_type

Add get_typeid_for_class helper

Add assertion for large types

Fix lttng map

Use helper methods to write into bulk type event logger buffer

Fix typo ep_rt_mono_send_method_details_event definition

Change s_name to ep_char8_t and write with gunichar2

mdh1418 commented on src/coreclr/scripts/genLttngProvider.py:
```diff
@@ -88,6 +88,34 @@
         "win:Binary"        :"const BYTE"
         }
 
+monoLttngDataTypeMapping ={
+        #constructed types
+        "win:null"          :" ",
+        "win:Int64"         :"const int64_t",
+        "win:ULong"         :"const uint32_t",
+        "win:count"         :"*",
+        "win:Struct"        :"const uint8_t *",
+        #actual spec
+        "win:GUID"          :"const int32_t",
+        "win:AnsiString"    :"const uint8_t*",
+        "win:UnicodeString" :"const uint8_t*",
+        "win:Double"        :"const double",
+        "win:Int32"         :"const int32_t",
+        "win:Boolean"       :"const bool",
+        "win:UInt64"        :"const uint64_t",
+        "win:UInt32"        :"const uint32_t",
+        "win:UInt16"        :"const uint16_t",
+        "win:UInt8"         :"const uint8_t",
+        "win:Pointer"       :"const size_t",
+        "win:Binary"        :"const BYTE"
+        }
```
I haven't tried it yet. I can try building and running the trace and pgo tools in coreCLR with this mapping and see how that works out. We diverged in https://github.com/dotnet/runtime/blob/ac6e77b676068c13878aa1a89bdccfaabde50d35/src/coreclr/scripts/genEventing.py#L44-L106 as well, should we try to unify that into one mapping as well?

josalem replied:
If everything can use one mapping, that would be great! I'm content with having two different lists if we need it, but reducing duplication is worthy effort.

lateralusX replied:
Original kept separate to make sure CoreCLR could continue to use its regular native types, while Mono keep use of C99 types. Some of the CoreCLR types won't work on Mono, so then we would standardize on the Mono types if unified, but since the mapping lives side by side, not sure how much it will benefit to unify them and they probably won't change ones established (only if we add new types into the etw manifest file)

mdh1418 replied:
I attempted to unify both mappings in genEventing and genLttngProvider, but `ep_char8_t` is causing problems. Perhaps unification can be done in a separate PR

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+// Estimate of how many bytes we can squeeze in the event data for the value struct
+// array.  (Intentionally overestimate the size of the non-array parts to keep it safe.)
+static const uint32_t k_max_bytes_type_values = (cb_max_etw_event - 0x30);
```
I took it from CoreCLR's implementation https://github.com/dotnet/runtime/blob/ac6e77b676068c13878aa1a89bdccfaabde50d35/src/coreclr/vm/eventtracepriv.h#L256, and that git blame has been squashed. I'm not sure what the non-array parts look like and why 48 bytes is an overestimate.

josalem replied:
Ah, I get it now. I didn't realize this was a translation of existing code in CoreCLR. 

For better or for worse, this appears to just be a heuristic born from back-of-the-envelope math. A comment to that effect is probably fine.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer[65536];
```
I'll add it as a const along with the other globals like `cb_max_etw_event` `k_max_bytes_type_values` and `k_max_count_type_values`. I also changed the buffer to be just `uint8_t *m_p_bulk_type_event_buffer` and allocating that memory size `p_type_logger->m_p_bulk_type_event_buffer = malloc(sizeof(uint8_t) * 65536);`. I think that is more correct in terms of filling the buffer when `ep_rt_mono_fire_bulk_type_event`. 

The original form was a mistake before knowing to try to allocate memory. With the memcpy format, it would have tried to save fixed_sized_data bytes into the first 4 indices, and then jump to the 33rd and 34th index to fill in the type's string name and so forth.

I also finally discovered the bug I had been looking for the past few days where fixed_sized_data is 25 bytes (typeID - 8, moduleID - 8, typeNameID - 4, Flags - 4, CorElementType - 1), and the dotnet-pgo tool understands it as that size; however, that struct will have a sizeof `32` from padding, leading to mismatch in actual offset and expected offset of the data.

mdh1418 replied:
Added the constant `k_size_of_event_buffer` (naming attributed to CoreCLR's `kSizeOfEventBuffer`)

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+{
+	if (p_type_logger->m_n_bulk_type_value_count == 0)
+		return;
+	memset(p_type_logger->m_p_bulk_type_event_buffer, 0, 65536 * sizeof(uint8_t));
```
I dont think the form has been adapted to Mono coding style yet :)
The structure of this function is almost line for line what is in CoreCLR https://github.com/dotnet/runtime/blob/ac6e77b676068c13878aa1a89bdccfaabde50d35/src/coreclr/vm/eventtrace.cpp#L1641-L1708. The only functional difference being that the bulk type event buffer is cleared out (I wanted to ensure that no leftover data from previous bulk type event firing lingers, but perhaps its not necessary) which is this memset line.

From what I understand, the overall functionality of this method is to fill in the struct's buffer with each of the values in the struct's Bulk Type Value array and fire the bulk type event, I'm not sure what parts feel orthogonal to that. I'm not quite certain, but do you mean splitting the function into smaller helper functions to make the logic more separated and independent? For example, 4 helper functions corresponding to `Copying fixed-size data`, `Copy type string name`, `Copy type parameter count`, `Copy type parameter array` in the CoreCLR's function's inline comments?

josalem replied:
I didn't realize this was a translation of CoreCLR code. This makes thing make more sense. In the CoreCLR case, this is an instance method and it is translated to C here.

I don't think a refactor is necessary but would make the function easier to grok. I will defer to your preference on this one. If there are multiple places where the same mutation takes place it might be nice to have a helper.

As for what to make into "helper" functions, I was imagining something akin to how @lateralusX structured most of the EventPipe code, e.g.,

```c
typedef struct {
    uint8_t buf[100];
    // ...
} MyStruct;

static
inline
void
ep_my_struct_init (MyStruct *myStruct) {...}

// ...

static
inline
void
ep_my_struct_reset (MyStruct *myStruct)
{
    memset (&myStruct->buf, 0, sizeof (MyStruct.buf));
}
```

If this is a translation of the CoreCLR code, we should add a comment explaining which code comes from there so that it is easier to trace its history.

mdh1418 replied:
I added comments that tie back to CoreCLR's code, also added init and fini for the BulkTypeEventLogger and an init for the BulkTypeValue structs.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
```
Thanks for the catch :). I haven't gotten around to trying to implement https://github.com/dotnet/runtime/blob/ac6e77b676068c13878aa1a89bdccfaabde50d35/src/coreclr/vm/eventtrace.cpp#L1880-L1896. As far as I can tell, thats the only instance where s_name would have been set, so this `else` actually isn't hit at all until s_name is set to something. 

I'll change the calculation to be `sizeof(wchar_t)`.

Note: I was somehow getting a `target->s_name` equivalent to `""` instead of it staying as `NULL`. so there are some additions to address that, such as 
```
	uint32_t s_name_len = 0;
	if (bulk_type_value->s_name)
		s_name_len = strlen(bulk_type_value->s_name);
```
and checking the length `if (!wsz_name || strlen(wsz_name) == 0)`

mdh1418 replied:
Changed the calculation of `ep_rt_mono_get_byte_count_in_event` to use `wchar_t`, also I realized that `s_name`, though never being set yet because I haven't implemented the snippet above, may contain junk from the g_malloc. Addressed that in the BulkTypeValue init `ep_rt_bulk_type_value_init`

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			   sizeof(target->fixed_sized_data));
+		i_size += sizeof(target->fixed_sized_data);
+
+		char *wsz_name = target->s_name;
```
Right, I naively set it to char as it is [`StackSString`](https://github.com/dotnet/runtime/blob/0e1f46b395052aa917aae377f5cc7f8a5eb07555/src/coreclr/vm/eventtracepriv.h#L237) and [`LPCWSTR`](https://github.com/dotnet/runtime/blob/0e1f46b395052aa917aae377f5cc7f8a5eb07555/src/coreclr/vm/eventtrace.cpp#L1673) in CoreCLR, and at first glance, I didn't know the 100% correct type for that in c.

josalem replied:
I'm not sure about `StackSString` (I'm guessing it's also wide char), but `LPCWSTR` is shorthand for `Long Pointer Const Wide-char STRing`. The CoreCLR code base is littered with macros like that. In general, `LPC` just means a `const pointer` to whatever the end type is.

mdh1418 replied:
I'm starting to understand the coreclr types more. [`StackSString`](https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.h#L953) seems to be [`InlineSString<512>`](https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.h#L807-L808) which seems to function like a byte array of size 512. More implementation details https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sstring.inl#L72-L99 and https://github.com/dotnet/runtime/blob/4cd4cff89cce3d48b70fdb3bfd604d3982d5e808/src/coreclr/inc/sbuffer.inl#L20-L45.

lateralusX replied:
Yes, it uses a preallocated buffer of 512 bytes, but since it inherits from SString it can also switch to dynamic allocated memory if it runs out of preallocated space. If you want to do something similar you can have a look at  EventPipeWriteEvent* functions that implement something similar, but if you want to reuse it, move it into a struct type, with a preallocated buffer, a pointer to the current buffer and a bool indicating if buffer pointed to is preallocated or dynamic allocated so it can be released, something like:

struct EventPipeStackString {
    ep_char8_t buffer[x];
    ep_char_t * ptr;
    uint32_t size;
    bool free_buffer;
};

and then a couple of functions like:

ep_stack_string_set_capacity (uint32_t size);
ep_stack_string_resize (uint32_t size);
ep_stack_string_get_ptr (uint32_t ensure_size);

and

ep_stack_string_init (EventPipeStackString *instance)
{
    instance->ptr = buffer;
    instance->size = x;
    instance->free_buffer = false;
}

ep_stack_string_fini (EventPipeStackString *instance)
{
    if (instance->free_buffer)
        ep_rt_free_utf8_string (instance->ptr)
    instance->ptr = NULL;
}

You could wrap the definitions above in some macros so you could create some preallocated buffers on some different sizes, 32, 64, 128, 256, 512 for example.

The implementation of ep_stack_string_get_ptr would make sure the buffer is of at least the requested size, switching to dynamic memory if needed.

Usage of the buffer would be to call ep_stack_string_get_ptr with the expected size and then use that pointer for the specific operation.

If you just need the pointer (without need to ensure it has a specific size, like doing strlen, you could implement a getter).

lambdageek replied:
Hey folks this is getting really elaborate:

1. Is the string supposed to be UTF8 or UTF16 when it goes into the buffer?  all this talk of `wchar_t` is confusing.  We generally don't use `wchar_t` in Mono - we use `gunichar2` for utf16 and `ep_char8_t*` for UTF8
2. Is it important to use fancy string storage here? Anything wrong with just heap allocating a buffer?
3. If we need to use fast fancy storage, how about using a MonoMempool ?

/cc @lateralusX

lateralusX replied:
Above was an example on how StackSString works and how it **could** be implemented on Mono, as an alternative to use regular stack based fixed buffer or heap allocated buffer. If we are not going to use the type logger outside of send method details events, then the pressure on the heap will be less, so for that use case heap allocated buffer might be enough. In CoreCLR the type logger is used for other events as well (like all GC events), so I mainly assume they use StackSString to reduce pressure on heap due to that.

Using a mempool instance during the lifetime of the logger could probably be an alternative as well, since we need to keep the buffers alive until emitting the types bulked by the logger, so all could be freed at the same time.

In EventPipe we use ep_char8_t for UTF8 and ep_char16_t for UTF16.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
+			memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size, wsz_name, name_size);
+			i_size += strlen(wsz_name);
```
Ah, the `wsz_name` is a typo, it should use `name_size`.
Things mainly weren't breaking since s_name isn't ever set. I haven't fully understood the snippet in https://github.com/dotnet/runtime/pull/68571#discussion_r863343068 to know if we need to do the same on mono yet.

mdh1418 replied:
Changed to name_size

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+		uint32_t *ptr_int = (uint32_t*)(p_type_logger->m_p_bulk_type_event_buffer + i_size);
+		*ptr_int = target->c_type_parameters;
+		i_size += 4;
```
Thanks! I can change it to that or even maybe `sizeof (*ptr_int)` if that is preferred to maintain the logical connection.

josalem replied:
`sizeof (*ptr_int)` is less readable than `sizeof (uint32_t)` but would also work.

mdh1418 replied:
Changed to `sizeof (uint32_t)`, thanks!

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+//---------------------------------------------------------------------------------------
+//
+// get_typeid_for_type is responsible for obtaining the unique type identifier for a
+// particular MonoType. MonoTypes are structs that are not unique pointers. There
+// can be two different MonoTypes that both System.Thread or int32 or bool[]. There
```
From my understanding there can be multiple MonoTypes `mono_type_1` and `mono_type_2` that are both of `MonoTypeEnum` `MONO_TYPE_I4(int32)` (so essentially the same type) but they have completely different pointers. This is problematic as MethodDetails and BulkType events revolve around the concept of having a unique handle to a type (its TypeID) that can be referenced to the one instance where that type is logged. 

CoreCLR is able to implement this through TypeHandle, but that doesn't exist for mono.

Instead, to get a sole TypeID to represent all `int32` types that are encountered when logging information of methods, we leverage the fact that there is only one `MonoClass` representing `int32`, and all `MonoType` can connect back to that `MonoClass` through [`mono_class_from_mono_type_internal`](https://github.com/dotnet/runtime/blob/0e1f46b395052aa917aae377f5cc7f8a5eb07555/src/mono/mono/metadata/class.c#L2171-L2236). Consequentially, that sole `MonoClass` will have a sole `this_arg` field and a sole `_byval_arg` field that can then be used as the unique handle for a type.

The brevity in the comment may be confusing, how can it be best improved? Also would this be the right place to note mono implementation details? @lambdageek

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+//      * type_id - Unique type identifier of mono_type
+//
+// Return Value:
+//      Index into array of where this type got batched. -1 if there was a failure.
```
Ah, thanks!

mdh1418 replied:
Changed back from `uint32_t` to `int`

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+	// Clear out p_val before filling it out (array elements can get reused if there
+	// are enough types that we need to flush to multiple events).
+	memset(p_val->rg_type_parameters, 0, 32);
+	memset(p_val->rg_mono_type_parameters, 0, 32);
```
Thanks, I'll put a macro that ties in https://github.com/dotnet/runtime/pull/68571#discussion_r859972436 into the comment explaining why 32

mdh1418 replied:
Added this macro as `k_size_of_type_parameter_array`

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		if (method_inst->type_argc > 1024) // ETW has a limit for maximum event size. Do not log overly large method type argument sets
+			return;
+
+	BulkTypeEventLogger *p_type_logger = malloc(sizeof(BulkTypeEventLogger));
```
also in general mono should use `g_malloc`/`g_free` not `malloc`/`free`

mdh1418 replied:
Thanks, changed to `g_malloc` and added the `ep_rt_mono_bulk_type_logger_init` function.

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	MonoClass *klass = method->klass;
+	if (klass)
+	{
+		MonoType *method_mono_type = m_class_get_byval_arg (klass);
+		method_type_id = get_typeid_for_type (method_mono_type);
```
This pattern seems pretty common.  Maybe add `get_type_id_for_class` as a helper?

mdh1418 replied:
Added! I think it applies twice.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	memset(p_val->rg_mono_type_parameters, 0, 32);
+	if (p_val->s_name)
+		p_val->s_name = '\0';
+	p_val->c_type_parameters = 0;
```
Thanks, I added that function right under the struct definition :)

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			return -1;
+		}
+	}
+	if (p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)
```
Done

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		{
+			// This type is apparently so huge, it's too big to squeeze into an event, even
+			// if it were the only type batched in the whole event.  Bail
+			return -1;
```
Changed return type to int

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// Batch up this type.  This grabs useful info about the type, including any
+	// type parameters it may have, and sticks it in m_rg_bulk_type_values
+	uint32_t i_bulk_type_event_data = ep_rt_mono_log_single_type (p_type_logger, mono_type, type_id);
+	if (i_bulk_type_event_data == -1)
```
Changed type to `int`

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		method_inst = method_ctx->method_inst;
+
+	if (method_inst)
+		if (method_inst->type_argc > 1024) // ETW has a limit for maximum event size. Do not log overly large method type argument sets
```
Added `k_max_method_type_argument_count` const

lateralusX commented on src/coreclr/scripts/genLttngProvider.py:
```diff
@@ -88,6 +88,34 @@
+        "win:UInt16"        :"const uint16_t",
+        "win:UInt8"         :"const uint8_t",
+        "win:Pointer"       :"const size_t",
+        "win:Binary"        :"const BYTE"
```
A couple of things you need to map differently:

"win:AnsiString"    :"const char*",
"win:UnicodeString" :"const ep_char8_t*"

"win:Pointer"       :"const void*",
"win:Binary"        :"const uint8_t"

mdh1418 replied:
Thanks, done!

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+void
+ep_rt_mono_log_type_and_parameters_if_necessary (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// TODO Log the type if necessary
```
It looks like SendMethodDetailsEvent in CoreCLR always logs the type with `ETW::TypeSystemLog::kTypeLogBehaviorAlwaysLog` for the two times it calls `ETW::TypeSystemLog::LogTypeAndParametersIfNecessary` https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L6442 and https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L6468, being consumed at https://github.com/dotnet/runtime/blob/5992145db2cb57956ee444aa0f0c2f3f85ee3673/src/coreclr/vm/eventtrace.cpp#L3275-L3280 and the `TypeLogBehavior` is recursively passed through. Since this PR aims to primarily implement `ep_rt_mono_send_method_details_event` and `ep_rt_mono_fire_bulk_type_event`, where the latter is currently only called by the former, the hash map wouldn't effectively be used unless something else called `ep_rt_mono_log_type_and_parameters_if_necessary`, which I'm not certain we have plans for that yet. Would it be better to just implement the `if necessary` when its actually used so it can be tested properly?

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,514 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
```
There shouldn't be any uses of `wchar_t` in Mono. if this is supposed to be UTF16, then use `gunichar2`.

mdh1418 replied:
Thanks, it is now using `gunichar2` for writing to the bulk type event logger buffer, and converting `s_name` from `ep_char8_t` to `gunichar2` using `g_utf8_to_utf16 (target->s_name, -1, NULL, NULL, NULL)` and freeing the bytes copied as discussed

lateralusX replied:
In EventPipe we use ep_char8_t (UTF8) as much as possible. If this string only need to be UTF16 in the serialized event payload (that's how nettrace strings are defined), then the utf8 -> utf16 should already be done inside the EventPipeWriteEvent* generated functions, so you should be able to fire an event with a string as UTF8 and it will be converted by the serializer. For an example, look at FireEtwMethodDCEndVerbose_V1, all parameters to that function is ep_char8_t * and will automatically be converted before written into the payload buffer.

lateralusX replied:
For each EventPipe provider there are auto generated functions that does serialization into buffers for all FireEtw* generated functions, there you find functions to serialize most EP types. You could use that as inspiration to align this with how all the other FireEtw* serialize parameters into the payload.

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	bulk_type_value->c_type_parameters = 0;
+	memset(bulk_type_value->rg_type_parameters, 0, k_size_of_type_parameter_array);
+	memset(bulk_type_value->rg_mono_type_parameters, 0, k_size_of_type_parameter_array);
+	if (bulk_type_value->s_name)
+		bulk_type_value->s_name[0] = '\0';
```
I don't understand what this is supposed to do.

The first time that `ep_rt_bulk_type_value_init` is called, the `s_name` field is NULL, so nothing happens.

When you later call it to re-initialize the bulk type value after it's been fired, this apparently overwrites the old string value's data?

That doesn't seem like what you want.  If you are storing a pointer to the name that is in some other data structure, you just corrupted someone else's data.

If you're storing a string that you allocated yourself, then you should free it (see my earlier suggestion about a destroy function) and set `s_name` to null.   You don't know how much storage you allocated for `s_name`, so if you're hoping to reuse the buffer you don't have enough info.

Keep it simple: initialize `s_name` to null. always store a copy of the name.  destroy the bulk type value after it's been flushed and deallocate `s_name` and reset it to NULL.

mdh1418 replied:
Since we are keeping `s_name` null for now, removed all instances of setting the first byte to `\0`

lambdageek reviewed (commented):
It's coming together

The main remaining question is around the memory management of `s_name`.

Also some nits about `intptr_t` (it's by definition either 8 or 4 bytes, depending on the architecture. so it's not a great idea to use it as a wire format type)

mdh1418 committed:
Removed temporary debug traces

Set BulkTypeValue s_name to null for now

Fixed debug trace formatting for pointers

Fix type_name_id for types with no token

Remove write_event_buffer_intptr as the size is different on Windows x86

Rename ep_rt_bulk_type_value_init to ep_rt_bulk_type_value_clear

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+	if (bulk_type_value->s_name)
+		while (bulk_type_value->s_name [s_name_len])
+			s_name_len++;
```
Thanks, removed for now since we are leaving `s_name` to be NULL.

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		}
+		for (int i = 0; i < target->c_type_parameters; i++)
+		{
+			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "target->rg_mono_type_parameters[%lu]: %lu\n", i, target->rg_mono_type_parameters[i]);
```
Thanks, I removed most of the traces now as they were temporary for debugging, the one left in I am using `%p` for `(gpointer)MonoType*` and `intptr_t`

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+		for (int i = 0; i < target->c_type_parameters; i++)
+		{
+			i_size += write_event_buffer_intptr (target->rg_type_parameters[i], ptr, &ptr);
```
Removed `write_event_buffer_intptr` helper and made this change, thanks!

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+		for (int i = 0; i < target->c_type_parameters; i++)
+		{
+			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "target->rg_mono_type_parameters[%lu]: %lu\n", i, target->rg_mono_type_parameters[i]);
+			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "target->rg_type_parameters[%lu]: %lu\n", i, target->rg_type_parameters[i]);
```
Removed these temporary traces

lambdageek reviewed (approved):
Mono bits LGTM. @josalem  or @lateralusX any other issues in the generator scripts or in ep-rt-mono that you think need to be addressed in this PR?

mdh1418 commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,579 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	EP_ASSERT (p_type_logger->m_n_bulk_type_value_count < k_max_count_type_values);
+
+	BulkTypeValue *p_val = &p_type_logger->m_rg_bulk_type_values[p_type_logger->m_n_bulk_type_value_count];
+	ep_rt_bulk_type_value_init (p_val);
```
Since `s_name` is now `NULL`, no longer allocating or de-allocating memory for it, and renamed `ep_rt_bulk_type_value_init` to `ep_rt_bulk_type_value_clear` to keep it in line with CoreCLR's

mdh1418 committed:
Utilize macros for array size initialization

Cleanup formatting and style

Remove unused write_event_buffer_utf16_str

Add BulkTypeEventLogger pointer to function description

mdh1418 committed:
[mono] Mark mono_class_has_finalizer as MONO_COMPONENT_API

mdh1418 committed:
Set global static const to macros

Change ep_rt_mono_get_byte_count_in_event type

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+void
+ep_rt_bulk_type_value_clear (BulkTypeValue *bulk_type_value)
+{
+	bulk_type_value->fixed_sized_data.type_id = 0;
```
nit: you could do a memset on the sizeof(BulkTypeValue) instead of setting all individual fields to 0. That would also be more resistant if new fields are added, since they would be automatically cleared.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+static
+BulkTypeEventLogger*
+ep_rt_bulk_type_event_logger_init ()
```
If you are going to align with how EventPipe uses init/fini and alloc/free, this function allocates a new type on the heap so it should be named ep_rt_bul_type_event_logger_alloc.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+
+static
+void
+ep_rt_bulk_type_event_logger_fini (BulkTypeEventLogger *p_type_logger)
```
And this, ep_rt_bulk_type_event_logger_free.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	// batch or whether we need to flush
+	int cb_val = ep_rt_mono_get_byte_count_in_event (p_val);
+	if (cb_val > K_MAX_BYTES_TYPE_VALUES) {
+		cb_val = ep_rt_mono_get_byte_count_in_event (p_val);
```
Why is the same call and condition double?

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	uint32_t c_params = p_val->c_type_parameters;
+	intptr_t rg_type_parameters [K_SIZE_OF_TYPE_PARAMETER_ARRAY];
+	MonoType *rg_mono_type_parameters [K_SIZE_OF_TYPE_PARAMETER_ARRAY];
+	for (uint32_t i = 0; i < c_params; i++) {
```
memcpy?

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	BulkTypeValue m_rg_bulk_type_values [K_MAX_COUNT_TYPE_VALUES];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer;
```
If you are not depending on order of fields, then you should move this field just after m_rg_bulk_type_values to reduce padding.

lateralusX commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,600 @@ ep_rt_mono_write_event_ee_startup_start (void)
+	if (method_inst)
+		method_inst_parameter_types_count = method_inst->type_argc;
+
+	intptr_t method_inst_parameters_type_ids [K_MAX_METHOD_TYPE_ARGUMENT_COUNT];
```
This will bring in a rather large stack allocation (8K on 64-bit) in potential deep callstacks, not sure we should do anything about it now, just wanted to call it out.

mdh1418 replied:
Right, there were complaints on CI about `method_inst_parameter_types_count` being variable at compile time. I just did this to make it a constant, and technically, we accept that many parameters in the method inst `if (method_inst && method_inst->type_argc > K_MAX_METHOD_TYPE_ARGUMENT_COUNT)`. Is there another way to keep it a constant at compile time? I'm now realizing that the type on CoreCLR is a `StackSArray`, which if its like the `StackSString` type, it will be a fixed size until it needs to allocate more memory.

Would we prefer to implement that sort of type?

lateralusX replied:
One option for this is to use g_alloca, dynamic stack allocation, that way you can alloc needed bytes on stack, just make sure you don't allocate to much (keep the limit of max method) and that you don't do it in any loops, alloca is not scoped on all platforms so doing it loops will just increase stack usage util function exists. We already use g_alloca in other parts of the runtime.

mdh1418 replied:
As discussed separately, we are now using a dedicated mempool for the BulkTypeEventLogger to assist in dynamically allocating enough memory for the relevant type parameters.

lateralusX reviewed (approved):
Overall it looks good, nice work! A couple of smaller things noted. I also think you should do a pass and drop all the variable prefixes used (p_, s_ m_ rg_ etc), its not Mono code style using that, and I assume you pretty much took variable names from CoreCLR C++ implementation and used them in this PR.

mdh1418 committed:
Address feedback

Drop variable prefixes

Clear BulkTypeValue via memset

Rename bulk_type_event_logger helpers to alloc and free

Remove redundant bulk type value byte count calculation

Utilize memcpy to copy type parameter arrays

Reduce padding by moving bulk_type_event_buffer field in struct

mdh1418 committed:
Change Event struct parameter _ElementSize type to size_t

mdh1418 committed:
Remove parallel type id array to reduce bookkeeping and create type ids from mono types when needed

Change type id type from intptr_t to uint64_t as ids in eventpipe are standardized to uint64_t

mdh1418 committed:
[mono][eventpipe] Add dedicated mempool for BulkTypeEventLogger to dynamically allocate enough space for type parameters

lambdageek commented on src/mono/mono/eventpipe/ep-rt-mono.c:
```diff
@@ -2857,6 +2857,585 @@ ep_rt_mono_write_event_ee_startup_start (void)
+// are no type parameters to log, but let's overestimate at 4 per value).
+#define K_MAX_COUNT_TYPE_VALUES ((uint32_t)(128 - 2) / 4)
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue bulk_type_values [K_MAX_COUNT_TYPE_VALUES];
+	uint8_t *bulk_type_event_buffer;
+	uint32_t bulk_type_value_count;
+	uint32_t bulk_type_value_byte_count;
+};
```
I would just add the mempool as an extra member here since its lifetime is tied to the BulkTypeEventLogger.  Then you don't have to pass the mempool as an extra argument to `ep_rt_mono_log_type_and_parameters` and other functions.

mdh1418 replied:
Moved the dedicated memory pool to be a BulkTypeEventLogger field! Thanks

lambdageek reviewed (commented):
small quality of life suggestion

mdh1418 committed:
Revert "Change Event struct parameter _ElementSize type to size_t"

This reverts commit 87f87f2f93457bed292d979668dbad01beb1e5f1.

mdh1418 committed:
Fix write_event_buffer parameter type to avoid C4267 error on windows

mdh1418 committed:
Move mem_pool under BulkTypeEventLogger

josalem reviewed (approved):
👍  LGTM

mdh1418 commented:
The failures look unrelated to this PR

mdh1418 commented:
Thanks @lambdageek @lateralusX @josalem for all of your help!
